src/interface/control_scrollbar.hpp

```diff
@@ -22,48 +23,67 @@
 #include "math/rectf.hpp"
 #include "math/vector.hpp"
 
-class DrawingContext;
-union SDL_Event;
-
-/** A generic template for a scrollbar */
+/** A scrollbar, which manages a scroll progress variable. */
 class ControlScrollbar final : public Widget
 {
 public:
-  ControlScrollbar();
+  ControlScrollbar(float total_region, float covered_region,
+                   float& progress, float mouse_wheel_speed = 10.f);
 
   virtual void draw(DrawingContext& context) override;
   virtual void update(float dt_sec) override;
 
   virtual bool on_mouse_button_up(const SDL_MouseButtonEvent& button) override;
   virtual bool on_mouse_button_down(const SDL_MouseButtonEvent& button) override;
   virtual bool on_mouse_motion(const SDL_MouseMotionEvent& motion) override;
+  virtual bool on_mouse_wheel(const SDL_MouseWheelEvent& wheel) override;
+
+  bool is_active() const;
+
+  void scroll(float amount);
+
+  void set_mouse_wheel_speed(float speed) { m_mouse_wheel_speed = speed; }
+
+  void set_rect(const Rectf& rect) { m_rect = rect; }
+  const Rectf& get_rect() const { return m_rect; }
+
+  float get_covered_region() const { return m_covered_region; }
+  float get_total_region() const { return m_total_region; }
+
+  void set_covered_region(float region);
+  void set_total_region(float region);
 
 private:
+  void check_regions();
+
+  Rectf get_bar_rect() const;
+
+private:
+  /** Determines the scroll speed, when using the mouse wheel. */
+  float m_mouse_wheel_speed;
+
   /** Whether or not the mouse is clicking on the bar */
   bool m_scrolling;
 
   /** Whether or not the mouse hovers above the bar */
   bool m_hovering;
 
   /** The length (height) of the region to scroll */
-  int m_total_region;
+  float m_total_region;
 
   /** The length (height) of the viewport for the region */
-  int m_covered_region;
+  float m_covered_region;
 
   /** The length (height) between the beginning of the viewport and the beginning of the region */
-  int m_progress;
-  
+  float& m_progress;
```

Comment:
Why a float reference?

Reply:
As the description says, it manages a scroll progress variable. That variable is an outside one, usually owned by the object which contains the scrollbar.

Reply:
Ah. Nice